sdk/core/core-rest-pipeline/src/policies/bearerTokenAuthenticationPolicy.ts

```diff
@@ -32,23 +142,130 @@ export interface BearerTokenAuthenticationPolicyOptions {
+    ...challengeCallbacks
+  };
+
+  /**
+   * We will retrieve the challenge only if the response status code was 401,
+   * and if the response contained the header "WWW-Authenticate" with a non-empty value.
+   */
+  function getChallenge(response: PipelineResponse): string | undefined {
+    const challenge = response.headers.get("WWW-Authenticate");
+    if (response.status === 401 && challenge) {
+      return challenge;
+    }
+    return;
+  }
 
   // This function encapsulates the entire process of reliably retrieving the token
   // The options are left out of the public API until there's demand to configure this.
   // Remember to extend `BearerTokenAuthenticationPolicyOptions` with `TokenCyclerOptions`
   // in order to pass through the `options` object.
-  const getToken = createTokenCycler(credential, scopes /* , options */);
+  const cycler = createTokenCycler(credential /* , options */);
 
   return {
     name: bearerTokenAuthenticationPolicyName,
+    /**
+     * If there's no challenge parameter:
+     * - It will try to retrieve the token using the cache, or the credential's getToken.
+     * - Then it will try the next policy with or without the retrieved token.
+     *
+     * It uses the challenge parameters to:
+     * - Skip a first attempt to get the token from the credential if there's no cached token,
+     *   since it expects the token to be retrievable only after the challenge.
+     * - Prepare the outgoing request if the `prepareRequest` method has been provided.
+     * - Send an initial request to receive the challenge if it fails.
+     * - Process a challenge if the response contains it.
+     * - Retrieve a token with the challenge information, then re-send the request.
+     */
     async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
-      const { token } = await getToken({
-        abortSignal: request.abortSignal,
-        tracingOptions: request.tracingOptions
-      });
-      request.headers.set("Authorization", `Bearer ${token}`);
-      return next(request);
+      // Allows users to easily set the authorization header.
+      function setAuthorizationHeader(accessToken: AccessToken) {
+        request.headers.set("Authorization", `Bearer ${accessToken.token}`);
+      }
+
+      if (callbacks?.authenticateRequest) {
+        await callbacks.authenticateRequest({
+          scopes,
+          request,
+          previousToken: cycler.cachedToken,
+          getToken: cycler.getToken,
+          setAuthorizationHeader
+        });
+      }
+
+      let response: PipelineResponse;
+      try {
+        response = await next(request);
+      } catch (err) {
+        response = err.response;
+      }
+      const challenge = getChallenge(response);
+
+      if (challenge && callbacks?.authenticateRequestOnChallenge) {
+        // processes challenge
+        const sendRequest = await callbacks.authenticateRequestOnChallenge(challenge, {
```

Comment:
TODO: Call it "shouldSendRequest" (Thanks, Deya!)

Reply:
Fixed here: https://github.com/Azure/azure-sdk-for-js/pull/13888/commits/5a4d88bedc5b759f3b1e86413b42556f1d47f5a9